OAuth로그인 및 회원가입 구현

dlehdhun opened the pull request:
## #️⃣ 연관된 이슈
- #31 

## 📝 작업 내용
- Passport 기반 OAuth 로그인 및 회원가입 기능 구현
- 카카오/구글 OAuth 연동
- 기존 회원은 로그인 처리, 신규 회원은 자동 회원가입되도록 로직 추가

dlehdhun committed:
feat: (#31) 카카오 OAuth 로그인 및 회원가입 구현

dlehdhun committed:
feat: (#31) 구글 OAuth 로그인 및 회원가입 구현

dlehdhun committed:
refactor: (#31) 구글 및 카카오 OAuth 인증 로직 통합

- AuthService 내의 중복된 로그인/회원가입 로직을 하나로 통합
- UserService 및 UserRepository 내의 중복된 회원가입 로직을 하나로 통합

dlehdhun committed:
refactor: (#31) 린트 오류 해결 및 타입 안정성 개선

rrgks6221 commented on src/auth/auth.controller.ts:
```diff
@@ -72,6 +76,46 @@ export class AuthController {
+
+    this.setRefreshToken(res, refreshToken);
+
+    return { accessToken };
```
제가 듣기로는 리다이렉트된 API 응답은 프론트엔드에서 파싱할 수 없다고 들은거같은데 이 부분 프론트엔드분과 확인 한번 해보시면 좋을거같아요

dlehdhun replied:
브라우저에서 http://localhost:3000/auth/kakao 로 직접 로그인 테스트했을 때는 토큰이 정상적으로 발급됐는데,
프론트와 연동하면 동작 방식이 달라질 수 있나요?

rrgks6221 commented on src/users/users.service.ts:
```diff
@@ -21,4 +23,19 @@ export class UsersService {
+  }
+
+  // 소셜 로그인 유저 조회
+  async findSocialUser(nickname: string): Promise<User | null> {
```
```suggestion
  async findByNickname(nickname: string): Promise<User | null> {
```

rrgks6221 commented on src/auth/auth.controller.ts:
```diff
@@ -72,6 +76,46 @@ export class AuthController {
+  }
+
+  @Get('kakao/callback')
+  @UseGuards(KakaoGuard)
```
이렇게 해야 정상적으로 req.user에 값이 들어갈거같은데 혹시 테스트가 됐을까요??

WEB에서 리다이렉트를 시나리오할 수 있는 파일이 없어서 아마 테스트가 안됐을거같은데 AI를 이용해서 간단한 html을 생성해 테스트해보시는것도 좋을거같아요

```suggestion
  @UseGuards(AuthGuard('kakao'))
```

dlehdhun replied:
`export class KakaoGuard extends AuthGuard('kakao')`로 선언해두었는데,`@UseGuards(AuthGuard('kakao'))`를 사용하는게 맞나요? 이것도 위에 처럼 http://localhost:3000/auth/kakao 로 테스트 할 때는 문제가 없었는데...

rrgks6221 commented on src/auth/strategies/google.strategy.ts:
```diff
@@ -0,0 +1,35 @@
+  validate(accessToken: string, refreshToken: string, profile: Profile) {
+    const { id, displayName } = profile;
+
+    if (!id || !displayName) {
+      throw new BadRequestException(OAUTH_ERRORS_CODES.INVALID_OAUTH_PROFILE);
+    }
```
이 경우는 어떤 경우에 발생하는것을 예상하고 예외처리 하신걸까요??
open API에서 scope를 정확히 전달했을 때 발생하지 않을 에러같아서요
타입 에러때문이라면 코드는 그대로 냅두되 에러 발생 원인은 서버 측이기때문에 500처리하는건 어떨까요??

dlehdhun replied:
카카오 측에서 필수 데이터(id, displayName)가 누락되는 것을 방지하고자 해당 검증을 추가했습니다. 원래는 500번때로 처리하려고 했는데 작업하면서 실수로 BadRequestException(400)로 넣어버렸네요... 500으로 수정하겠습니다!

rrgks6221 reviewed (commented):
kakao랑 google 작업을 나눴다면 더 빠르게 작업 후 피드백이 됐을거같아요!
대부분의 코멘트가 kakao, google 둘 다 해당되기도 하구요

dlehdhun committed:
docs: (#39) OAuth 로그인/회원가입 Swagger 문서 추가

kimjiwonpg98 commented on src/auth/auth.service.ts:
```diff
@@ -90,4 +91,37 @@ export class AuthService {
+  // 소셜 로그인
+  async socialLogin(socialUser: SocialLoginPayload): Promise<TokenDto> {
+    const userNickname = await this.userService.findSocialUser(socialUser.nickname);
+    if (userNickname && userNickname.nickname !== socialUser.nickname) {
```
```suggestion
    if (userNickname) {
```
위에서 nickname으로 찾았는데 nickname이 맞지 않을 때를 조건에 넣은게 맞는지 확인 부탁드려요.

개인적인 생각으로는 oauth에서의 닉네임은 충분히 겹칠 수 있어서 유니크값으로 닉네임을 운영할거라면 서비스만의 닉네임을 따로 받는게 좋아보이긴합니다.

dlehdhun replied:
감사합니다! 저도 처음에는 nickname에 unique키를 설정해두었기 때문에 중복 여부까지 확인했었습니다. 그런데 닉네임이 겹칠 확률이 높다고 판단하여 nickname의 unique 제약을 제거하기로 했습니다. 이부분은 #42 이슈에서 해결할 예정입니다!

dlehdhun committed:
refactor: (#42) username 유니크 키 제거 및 중복 검증 로직 제거

- User 테이블 username 컬럼의 UNIQUE 키 삭제
- username관련 중복 검증 로직 제거
- 타입명 userdata → UserData로 네이밍 수정

dlehdhun committed:
refactor: (#31) 메서드 네이밍 수정 및 예외 처리 변경

- findSocialUser -> findByNickname으로 네이밍 변경
- 로직 실패 시 BadRequest(400) 대신 InternalServerError(500)를 반환하도록 수정

dlehdhun committed:
Merge pull request #46 from modern-agile-team/docs/#39/oauth-auth-swagger

OAuth 로그인/회원가입 Swagger 문서 추가

dlehdhun committed:
Merge branch 'feat/#31/oauth-auth' into fix/#42/handle-duplicate-nickname

dlehdhun committed:
Merge pull request #45 from modern-agile-team/fix/#42/handle-duplicate-nickname

nickname 유니크 키 제거 및 중복 검증 로직 제거

dlehdhun committed:
refactor: (#31) 린트 오류 해결
